### PR TITLE
feat: support frontend plugins via env.config.jsx

### PR DESCRIPTION
**Pull request by arbrandes**
This provides a mechanism to configure frontend plugins using a new PLUGIN_SLOTS filter and a series of patches that modify a base `env.config.jsx`, in such a way that multiple plugins can take advantage of the file (including for purposes beyond frontend plugins) without clobbering each other.

(This is the third take.  Prior: #233, #234)

### CODE REVIEW

**Comment by arbrandes** [comment]
@regisb, how about this?

File: README.rst
Comment (regisb) [inline README.rst:327]: This is a bit of a nitpick: I suggest having separate items for every plugin slot. I.e:

```
 PLUGIN_SLOTS.add_items(
    ("all", "footer_slot",
        """
          {
            /* Hide the default footer. */
            op: PLUGIN_OPERATIONS.Hide,
            widgetId: 'default_contents'
          }),
    ("all", "footer_slot", """..."""),
    ...
)
```

In particular, with this approach we don't have to be too careful with trailing commas.

This is a minor concern and should not be blocking this PR.

File: tutormfe/plugin.py
Comment (regisb) [inline tutormfe/plugin.py:92]: Make sure to use [tutor.hooks.lru_cache](https://docs.tutor.edly.io/reference/api/hooks/index.html#tutor.hooks.lru_cache) (and fix get_mfes above)

File: tutormfe/plugin.py
Comment (regisb) [inline tutormfe/plugin.py:97]: `PLUGIN_SLOTS.iterate()` is syntactic sugar for `PLUGIN_SLOTS.apply([])`.

File: tutormfe/templates/mfe/build/mfe/env.config.jsx
Comment (regisb) [inline tutormfe/templates/mfe/build/mfe/env.config.jsx:3]: Having two patch statements in quick succession is a clue that we should remove one. If some patches need to be at the bottom, then they can be added with a lower priority.

File: tutormfe/plugin.py
Comment (regisb) [inline tutormfe/plugin.py:121]: We are never using `mfe_name` in the templates. Does that mean that we should simply not return it?

File: tutormfe/templates/mfe/build/mfe/env.config.jsx
Comment (regisb) [inline tutormfe/templates/mfe/build/mfe/env.config.jsx:42]: nitpick: The "-tail" and "-head" suffixes are not used for other patch statements in Tutor. For instance, the patches in openedx/Dockerfile use "-pre", "-post", "-final". I'm not sure which convention is best, but we should probably follow the existing one.

**Review by regisb — COMMENTED** [review]
I think this is a great improvement, and I only have minor comments. Looking forward to merging this.

File: tutormfe/plugin.py
Comment (arbrandes) [reply tutormfe/plugin.py:92]: Got it, will do!

File: tutormfe/templates/mfe/build/mfe/env.config.jsx
Comment (regisb) [reply tutormfe/templates/mfe/build/mfe/env.config.jsx:3]: OK let's keep those patches then.

**Comment by arbrandes** [comment]
@regisb, ready for a second round.

File: tutormfe/templates/mfe/build/mfe/env.config.jsx
Comment (regisb) [inline tutormfe/templates/mfe/build/mfe/env.config.jsx:4]: Is this idiomatic? Shouldn't it be `if(!(slot_name in config.pluginSlots))`? Or `if(config.pluginSlots[slot_name] === null)`?

**Review by regisb — APPROVED** [review]
Stellar!

File: tutormfe/templates/mfe/build/mfe/env.config.jsx
Comment (hinakhadim) [inline tutormfe/templates/mfe/build/mfe/env.config.jsx:32]: Is there any particular reason for adding plugin_config object one by one?

Could we pass array for one plugin_slot (keeping in mind multiple plugins can have same slots)?
image

Like this:
```
PLUGIN_SLOTS.add_items([
        (
            mfe,
            "footer_slot",
            """
            [
    {
        op: PLUGIN_OPERATIONS.Hide,
        widgetId: 'default_contents',
    }
    {
      op: PLUGIN_OPERATIONS.Insert,
      widget: {
        id: 'default_contents',
        type: DIRECT_PLUGIN,
        priority: 1,
        RenderWidget: <Footer />,
      },
    },
    {
      op: PLUGIN_OPERATIONS.Insert,
      widget: {
        id: 'read_theme_cookie',
        type: DIRECT_PLUGIN,
        priority: 2,
        RenderWidget: AddDarkTheme,
      },
    },
  ]"""
        ),
    ])
```

File: tutormfe/templates/mfe/build/mfe/env.config.jsx
Comment (hinakhadim) [reply tutormfe/templates/mfe/build/mfe/env.config.jsx:4]: I think only `if (slot_name in config.pluginSlots)` is fine.

File: tutormfe/templates/mfe/build/mfe/env.config.jsx
Comment (hinakhadim) [reply tutormfe/templates/mfe/build/mfe/env.config.jsx:32]: Apologies, I just realized that objects are already passing in array. I need to pass without array brackets:
```javascript

PLUGIN_SLOTS.add_items([
        (
            mfe,
            "footer_slot",
            """
            
    {
        op: PLUGIN_OPERATIONS.Hide,
        widgetId: 'default_contents',
    },
    {
      op: PLUGIN_OPERATIONS.Insert,
      widget: {
        id: 'default_contents',
        type: DIRECT_PLUGIN,
        priority: 1,
        RenderWidget: <Footer />,
      },
    },
    {
      op: PLUGIN_OPERATIONS.Insert,
      widget: {
        id: 'read_theme_cookie',
        type: DIRECT_PLUGIN,
        priority: 2,
        RenderWidget: AddDarkTheme,
      },
    },
  """
        ),
    ])
```

File: README.rst
Comment (hinakhadim) [inline README.rst:373]: nit: targetting -> targeting

File: README.rst
Comment (hinakhadim) [inline README.rst:569]: Could we update the documentation to specify that if a component is exported as a default export, it should be accessed using Component.default when dynamically imported? This clarification will help ensure proper usage and avoid potential errors during runtime.

I was testing with indigo and here's my file:
```
// import Footer from '@edly-io/indigo-frontend-component-footer';

...

if (process.env.APP_ID == 'learner-dashboard') {
const Footer = await  import('@edly-io/indigo-frontend-component-footer');
...
{
      op: PLUGIN_OPERATIONS.Insert,
      widget: {
        id: 'default_contents',
        type: DIRECT_PLUGIN,
        priority: 1,
        RenderWidget: <Footer />,
      },
    },
...
}

```

It is throwing error `Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: object.` 

If I use `const { Footer } = await  import('@edly-io/indigo-frontend-component-footer');`, it throws this error `Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.`.

The footer is exported as default. `export default Footer;`. The errors are generated when I shift to dynamic import `await (...)`. Could we add in doc that we have to use `Footer(Component).default` in case of dynamic imports.

**Review by hinakhadim — APPROVED** [review]

File: tutormfe/templates/mfe/build/mfe/env.config.jsx
Comment (arbrandes) [reply tutormfe/templates/mfe/build/mfe/env.config.jsx:4]: On syntax: I guess it's a matter of opinion.  MDN [likes `=== false`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/in#try_it) for negating the `in` operator.  But I do thing `in` is the best check, here.  It's unambiguous as to whether the property is defined or not.

But we do have to use the negation, here.  It avoids duplicating code.

File: tutormfe/templates/mfe/build/mfe/env.config.jsx
Comment (arbrandes) [reply tutormfe/templates/mfe/build/mfe/env.config.jsx:16]: Sure.  But I don't know what the behavior is going to be if you actually call frontend-platform's getConfig while actually setting config.

In any case, I'll just call it `setConfig`, then.

File: tutormfe/templates/mfe/build/mfe/env.config.jsx
Comment (arbrandes) [reply tutormfe/templates/mfe/build/mfe/env.config.jsx:32]: There are different ways you can load the configuration in.  @regisb suggested we document it one-by-one to avoid issues with trailing commas.  Though in this latest iteration of `env.config.jsx` it doesn't really matter if there are trailing commas or not, I guess it keeps it simple.

Otherwise, yes you can totally use a single `add_item()` with a string that has multiple ops for one slot.

File: README.rst
Comment (arbrandes) [reply README.rst:569]: Sure, I'll point to the MDN documentation.  They mention that you have to use [the `default` key](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import#module_namespace_object).

Comment (arbrandes) [reply README.rst]: They actually document a better way [further down in the page](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import#importing_defaults), which is what I ended up putting up as an example.

**Comment by regisb** [comment]
@DawoudSheraz can you please merge and rebase the sumac branch on top of nightly?

**Comment by DawoudSheraz** [comment]
> I will test this branch locally with indigo nightly. Hina is on leaves this week, she mentioned that this can break indigo. I would verify that before merging.

I tested this locally and indigo's dark theme toggle is not working as expected. It is not appearing consistently across pages (for instance, it does not appear on the learner dashboard). Even when it appears, the toggle does not retain its value when navigating to different pages. Hina mentioned that this could impact indigo. 

I am ok with merging considering Hina will be back from leaves next week and can look into updating Indigo.

**Comment by arbrandes** [comment]
Let me know if I can help with anything.  We should probably try and get this (or something like this) merged this week so it comes out with Sumac.

**Comment by regisb** [comment]
@hinakhadim friendly nudge :)

**Comment by hinakhadim** [comment]
Hi @arbrandes ,

There's an issue generating with this PR on tutor-indigo that is it's not picking up by any MFE. The image building is successful and when we try to see in browser. The changes are not appearing. Steps:
- install indigo footer in leaner-dashboard MFE
- the contents of `env.config.jsx` file:
```
import React, { useEffect } from 'react';
import Cookies from 'universal-cookie';

import { getConfig } from '@edx/frontend-platform';

function addPlugins(config, slot_name, plugins) {
  // ...
}

let themeCookie = 'indigo-toggle-dark';
let themeCookieExpiry = 90; // days

const AddDarkTheme = () => {
  // ...

  return (<div />);
};

async function setConfig () {
  let config = {
    pluginSlots: {}
  };

  try {
    /* We can't assume FPF exists, as it's not declared as a dependency in all
     * MFEs, so we import it dynamically. In addition, for dynamic imports to
     * work with Webpack all of the code that actually uses the imported module
     * needs to be inside the `try{}` block.
     */
    const { DIRECT_PLUGIN, PLUGIN_OPERATIONS } = await import('@openedx/frontend-plugin-framework');

const Footer = await import(/* webpackIgnore: true */`@edly-io/indigo-frontend-component-footer`).catch(() => null); // Fallback to null if not available

    if (process.env.APP_ID == 'authn') {
    }
    if (process.env.APP_ID == 'authoring') {
    }
    
    if (process.env.APP_ID == 'learner-dashboard') {
      addPlugins(config, 'footer_slot', [ 
            {
                op: PLUGIN_OPERATIONS.Hide,
                widgetId: 'default_contents',
            },
            {
                op: PLUGIN_OPERATIONS.Insert,
                widget: {
                    id: 'default_contents',
                    type: DIRECT_PLUGIN,
                    priority: 1,
                    RenderWidget: Footer.default,
                },
            },
            {
            op: PLUGIN_OPERATIONS.Insert,
                widget: {
                    id: 'read_theme_cookie',
                    type: DIRECT_PLUGIN,
                    priority: 2,
                    RenderWidget: AddDarkTheme,
                },
            },
  ]);
    }
   
  } catch { }

  return config;
}

export default setConfig;
```

- I was thinking that `AddDarkTheme` func is not working. But the default footer is not hidden and new Indigo footer is not applied. 
- I added the previous `env.config.jsx` that exists in Indigo. and that works perfectly.


Reason of Issue: 
I'm assuming that this is causing because of `setConfig`. `export default setConfig` is a function and in the previous `env.config.jsx` (already used in indigo now), we are exporting object `let config = {...}; export default config;`


Could you please help in resolving it? I tried to figure out by calling `setConfig` but it didn't work. Because it's also `async` function and return Promise. Also, if you help how to add logger to `env.config.jsx`, then it helps in debugging it easily.

**Comment by arbrandes** [comment]
@hinakhadim, lemme try and reproduce it on my end.  Is there a PR to tutor-indigo that I can refer to?

**Comment by hinakhadim** [comment]
Sure, please. Here's the PR Link https://github.com/overhangio/tutor-indigo/pull/109

It might be `webpack Ignore` is causing this issue because I remember it was working before.  But the problem is `ora-grading` fails the build if I don't import Footer with `webpack Ignore`. (FYI: I also tried by importing Footer in `if` conditions. And one thing is that we aren't included footer in ora-grading but still I don't have idea why it's reading the Footer import during build and produces error). 

Error is in this conversation: https://github.com/overhangio/tutor-mfe/pull/240#issuecomment-2511793463 . At that time my env.config.jsx was like: (As you can see, footer in even imported in `learner-dashboard condition`. It has nothing with `ora-grading, but still failed the build`.
```
...
try{

const { DIRECT_PLUGIN, PLUGIN_OPERATIONS } = await import('@openedx/frontend-plugin-framework');
 
     if (process.env.APP_ID == 'authn') {
    }
    ...
    if (process.env.APP_ID == 'learner-dashboard') {
        const Footer = await import("@edly-io/indigo-frontend-component-footer");
         
        addPluginSlot(...);
    }
   ...
   if (process.env.APP_ID == 'ora-grading') {
    }
    ...
} catch {}
...
```


After adding `webpack Ignore` , image building is becoming successful but MFEs are not picking this file or may be picking but not applying any plugin slot.

**Comment by arbrandes** [comment]
@hinakhadim, I have some findings to discuss, but I'll do it as a review to https://github.com/overhangio/tutor-indigo/pull/109.

**Comment by arbrandes** [comment]
@hinakhadim, now that the MFEs have been fixed, let me know if things are good with https://github.com/overhangio/tutor-indigo/pull/109 so we can merge this.

**Comment by hinakhadim** [comment]
Things are working good. Just waiting for a heads-up from another member regarding testing before we proceed with the merge. Status: https://github.com/overhangio/tutor-indigo/pull/109#issuecomment-2523864661

**Comment by DawoudSheraz** [comment]
@arbrandes Hi. Please fix the failing tests so that this can be merged. Thanks

**Comment by arbrandes** [comment]
@DawoudSheraz, I'm looking into it, but to be honest, it'll probably take me a while.  My expertise with Python type annotations is limited.

**Comment by arbrandes** [comment]
Plus, I'm not currently set up to reproduce the test environment. 😓

File: tutormfe/plugin.py
Comment (regisb) [inline tutormfe/plugin.py:97]: This annotation should not be necessary, let's get rid of it.

**Comment by hinakhadim** [comment]
For the issue at line 124, I'm not entirely sure if this is the most appropriate solution, but it does resolve the problem
```
def get_mfe(mfe_name: str) -> MFE_ATTRS_TYPE:
    mfes : dict[str, MFE_ATTRS_TYPE] = get_mfes() 
    return mfes.get(mfe_name, {})
```

File: tutormfe/plugin.py
Comment (arbrandes) [reply tutormfe/plugin.py]: Yeah, this was just a dumb attempt to get the linting to stop complaining before I got python 3.9 set up locally. 😅

File: tutormfe/plugin.py
Comment (arbrandes) [reply tutormfe/plugin.py:97]: Nice catch!  This solved the linting cascade.

File: tutormfe/plugin.py
Comment (arbrandes) [inline tutormfe/plugin.py:122]: @hinakhadim, your previous comment got me on the right track, thanks - but I didn't want to change running code just for the sake of a type hint, so this is what I did.  It seems there's significant precedent for `Any` in the codebase, so I expect it to be fine.

**Comment by arbrandes** [comment]
Ok, finally got pyenv properly set up so I could reproduce the failures.  Thanks for the reviews!

File: tutormfe/plugin.py
Comment (regisb) [reply tutormfe/plugin.py:123]: The issue is caused by the lru_cache decorator, which is not correctly typed. You can bypass this issue without `Any` by writing:

```
def get_mfe(mfe_name: str) -> MFE_ATTRS_TYPE:
    mfe: MFE_ATTRS_TYPE = get_mfes().get(mfe_name, {})
    return mfe
```

**Pull request merged**
